### PR TITLE
fix: [CDS-86229]: Added support for parallel steps within container step group

### PR DESCRIPTION
**Pull request by ivan-83**
Added support for parallel steps within the container step group.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/57089)

### CODE REVIEW

**Comment by ivan-83** [comment]
trigger uts

**Comment by ivan-83** [comment]
trigger ut8

**Comment by sahilhindwani-harness** [comment]
*[`879-pipeline-ci-commons/src/main/java/io/harness/plancreator/steps/pluginstep/K8sStepGroupHandler.java` line 42 at r1](https://reviewable.io/reviews/harness/harness-core/57089#-NmdciqR2cPX4x_Prjft:-NmdciqSCi_gqETbmagW:bt6whtd) ([raw file](https://github.com/harness/harness-core/blob/338c47b5b532c8d2352a946b512fb73e9eb94c8a/879-pipeline-ci-commons/src/main/java/io/harness/plancreator/steps/pluginstep/K8sStepGroupHandler.java#L42)):*
> ```Java
>     String initNodeId = "init-" + ctx.getCurrentField().getNode().getUuid();
> 
>     String nextNodeId = getNextNodeId(stepsField);
> ```

Can we use this function? ```GenericPlanCreatorUtils.obtainNextSiblingField(currentField);```

**Comment by sahilhindwani-harness** [comment]
*[`879-pipeline-ci-commons/src/main/java/io/harness/plancreator/steps/pluginstep/K8sStepGroupHandler.java` line 42 at r1](https://reviewable.io/reviews/harness/harness-core/57089#-NmdciqR2cPX4x_Prjft:-Nmdcq6L8JEkGVfDnS5Y:b-a1djsc) ([raw file](https://github.com/harness/harness-core/blob/338c47b5b532c8d2352a946b512fb73e9eb94c8a/879-pipeline-ci-commons/src/main/java/io/harness/plancreator/steps/pluginstep/K8sStepGroupHandler.java#L42)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

Can we use this function? ```GenericPlanCreatorUtils.obtainNextSiblingField(currentField);```
</blockquote></details>

The above function will also take care if next is a normal step group inside step group.

File: 879-pipeline-ci-commons/src/main/java/io/harness/plancreator/steps/pluginstep/K8sStepGroupHandler.java [hunk 1]
```diff
@@ -38,10 +45,20 @@ public PlanNode handle(StepGroupElementConfig config, PlanCreationContext ctx, Y
...
-    String nextNodeId = stepsField.getNode().asArray().get(0).getField("step").getUuid();
-    ByteString advisorParametersInitStep = ByteString.copyFrom(
-        kryoSerializer.asBytes(OnSuccessAdviserParameters.builder().nextNodeId(nextNodeId).build()));
     return InitContainerV2StepPlanCreator.createPlanForField(
-        initNodeId, initContainerV2StepInfo, advisorParametersInitStep, "InitializeContainer");
+        initNodeId, initContainerV2StepInfo, getAdviserObtainment(stepsField), "InitializeContainer");
+  }
+
+  private AdviserObtainment getAdviserObtainment(YamlField stepsField) {
+    YamlField siblingField = GenericPlanCreatorUtils.obtainNextSiblingField(stepsField, Arrays.asList(STEP, PARALLEL));
```
Comment (ivan-83) [inline 879-pipeline-ci-commons/src/main/java/io/harness/plancreator/steps/pluginstep/K8sStepGroupHandler.java:53]: We are not supporting the nested step group within the container step group.

**Comment by ivan-83** [comment]
trigger uts

**Comment by sahilhindwani-harness** [comment]
*[`879-pipeline-ci-commons/src/main/java/io/harness/plancreator/steps/pluginstep/K8sStepGroupHandler.java` line 53 at r2](https://reviewable.io/reviews/harness/harness-core/57089#gh-1437018825:-NmfgooO7ieue-W8xDRr:byxj01b) ([raw file](https://github.com/harness/harness-core/blob/7812a344de1db35d3d552adad89f5be0ba0bc176/879-pipeline-ci-commons/src/main/java/io/harness/plancreator/steps/pluginstep/K8sStepGroupHandler.java#L53)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

We are not supporting the nested step group within the container step group. 
</blockquote></details>

Any reason for not supporting normal step groups inside container step groups?

File: 879-pipeline-ci-commons/src/main/java/io/harness/plancreator/steps/pluginstep/K8sStepGroupHandler.java (hunk 1 above)
Comment (ivan-83) [reply 879-pipeline-ci-commons/src/main/java/io/harness/plancreator/steps/pluginstep/K8sStepGroupHandler.java:53]: @sahilhindwani-harness can we discuss this point?

**Comment by ivan-83** [comment]
trigger uts

**Review by sahilhindwani-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/57089)* status: 4 file(s) to review. 1 of 2 approvals obtained.

**Comment by ivan-83** [comment]
trigger uts

**Comment by ivan-83** [comment]
trigger uts

**Comment by ivan-83** [comment]
trigger messagecheck

**Comment by ivan-83** [comment]
trigger messagecheck

**Comment by ivan-83** [comment]
trigger uts

**Review by sahilhindwani-harness — APPROVED** [review]
> Reviewed 1 of 3 files at r2, 2 of 2 files at r3, 1 of 1 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/57089)* status: 6 file(s) to review. 1 of 2 approvals obtained.

**Review by PragyeshMishra01 — APPROVED** [review]

**Comment by PragyeshMishra01** [comment]
trigger review

**Pull request merged**